{% extends 'base.html' %}
{% block css %}
<link
  href="{{ url_for('static', filename='css/tooltip.css') }}"
  rel="stylesheet"
/>
{% endblock %}
{% block content %}
<div class="row">
  <h2>Gene Mapping Via BSA(Bulked Segregant Analysis)</h2>
  <div class="panel-heading">
    <p>1. Wild Bulk and Mutant Bulk are <b>required</b>.</p>
    <p>2. Wild Parent and Mutant Parent are suggested if available.</p>
    <p>3. Background could help you filter out some variants, but you should be much careful the genetic
      background and phenotype of sample you selected.</p>
    <h3 class="panel-title">Please login and upload your vcf file if you not find the sample you want in the boxes
      below.</h3>
  </div></br>

  <div class="row">

    {% if pri_samples %}
    <div class="col-xs-3">
      <h3 class="panel-title">My samples</h3>
      <select name="from[]" id="multi_d" class="form-control multi_d" size="67" multiple="multiple">
        {% for sample in pri_samples %}
        <option value="private">{{ sample }}</option>
        {% endfor %}
      </select>
    </div>
    {% endif %}

    {% if pub_samples %}
    <div class="col-xs-3">
      <h3 class="panel-title">Public samples</h3>
      <select name="from[]" id="multi_d2" class="form-control multi_d" size="67" multiple="multiple">
        {% for sample in pub_samples %}
        <option value="public">{{ sample }}</option>
        {% endfor %}
      </select>
    </div>
    {% endif %}


    <div class="col-xs-2">
      <button type="button" id="multi_d_rightAll" class="btn btn-default btn-block" style="margin-top: 20px;"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>

      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>

      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>

      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>
      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>
    </div>

    <div class="col-xs-4">
      <b>Wild Bulk (Required)</b>
      <select name="to[]" id="multi_d_to" class="form-control" size="8" multiple="multiple"></select>

      <br />
      <hr /><br />

      <b>Mutant Bulk (Required)</b>
      <select name="to_2[]" id="multi_d_to_2" class="form-control" size="8" multiple="multiple"></select>
      <br />
      <hr /><br />

      <b>Wild Parent (Optional)</b>
      <select name="to_3[]" id="multi_d_to_3" class="form-control" size="8" multiple="multiple"></select>
      <br />
      <hr /><br />

      <b>Mutant Parent (Optional)</b>
      <select name="to_4[]" id="multi_d_to_4" class="form-control" size="8" multiple="multiple"></select>
      <br />
      <hr /><br />

      <b>Background (Optional)</b>
      <select name="to_5[]" id="multi_d_to_5" class="form-control" size="8" multiple="multiple"></select>
    </div>
  </div>
  <br /><br />
  <h3 class="panel-title"><b>Optional parameters [SNP Density/varBScore]:</b></h3>
  <br>

  <form>

    <div class="form-group row">
        <label for="min_depth" class="col-xs-5" style="font-weight:normal;">Minimum Read Depth
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="To filter out low coverage/depth variations. <b>Default 5</b>. Any variation coverage less than the setted value will be ignored."
            >?</span
          ></label>
        <div class="col-xs-7">
          <input id="min_depth" type="input" value="" placeholder="5" >
        </div>
    </div>

    <div class="form-group row">
      <label for="w-snp-num" class="col-xs-5" style="font-weight:normal;">SNP Window Size
        <span
        class="badge"
        data-toggle="tooltip"
        data-placement="top"
        data-html="true"
        title="Number of SNP in a slide window. <b>Default 20</b>. Suggest [5-30]. The algorithm will calculate the SNP linkage-coefficient in a window. A smaller value will have better resolution, and the more SNP in a window will eliminate more noise. Please try different parameters to obtain the best results. "
        >?</span
      >
    </label>
      <div class="col-xs-7">
        <input id="w-snp-num" type="input" value="" placeholder="20" >
      </div>
    </div>

    <div class="form-group row">
      <label for="s-snp-num" class="col-xs-5" style="font-weight:normal;">SNP Step Size
        <span
        class="badge"
        data-toggle="tooltip"
        data-placement="top"
        data-html="true"
        title="Number of SNP as slide step. <b>Default 5</b>. Suggest [5-20]. A smaller value will have better resolution but it will take more time to calculate. When your data from high generation population, it is feasible to increase the number of SNPs appropriately. The default parameter 5 work well for most data sets."
        >?</span
      >    
    </label>
      <div class="col-xs-7">
        <input id="s-snp-num" type="input" value="" placeholder="5" >
      </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Allele Frequency Filter [<b>Wild/Mutant</b>]
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="This parameter will filter out the SNPs with high heterozygosity rate. <b>Default 0.39 for F2; 0.3 for F3/DH/RIL</b>.
            </br>
            <b>F2</b>: The targets genotype of <b>dominant character bulk data from F2 population</b> are heterozygosity, and the candidate gene allele frequency in dominant character bulk expect as 1/3 (Aa 2:1). So, the parameter for F2 should set as 0.36/0.39/0.43 or any other value large than 1/3 and less than 0.45. If your bulk sample from F2 but validate by F3 which two bulk genotypes are expected homozygote. The parameter should set follow the rule as bellow.
            </br>
            <b>F3/RIL/DH</b> and any population genotype are expected homozygote in both tow bulks. The allele frequency is equal or very close to 0 or 1. The parameter could set as 0.1/0.2/0.3 or any other value less than 0.3.
            "
            >?</span
          >               
        </label>
        <div class="col-xs-7">
            <input id="ref_freq" type="input" value="" placeholder="0.3" >
        </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Allele Frequency Filter [<b>Wild/Mutant Parent</b>]
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="If the parent data were provided, this parameter will filter out the SNPs with high heterozygosity rate. Usually, most alleles of parentâ€™s genotype are homozygote, the allele frequency is equal or very close to 0 or 1. The value was suggested as <b>0.05 or 0.1</b> to avoid impact of sequencing error."
            >?</span
          >                
        </label>
        <div class="col-xs-7">
            <input id="p_ref_freq" type="input" value="" placeholder="0.3" >    
        </div>
    </div>    
    
    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Allele Frequency Filter [<b>Background</b>]
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="If you try to add dataset into background to help improve gene mapping, you should be much careful about the genetic background and phenotype of those sample. The expected allele frequency of samples in this bulk depends on what samples you selected. "
            >?</span
          >          
        </label>
        <div class="col-xs-7">
            <input id="background_ref_freq" type="input" value="" placeholder="0.3" >       
        </div>
    </div>  
  </form>
  <br />
  <h3 class="panel-title"><b>Optional parameters [QTLseqr/ED/G']:</b></h3>
  <br>
  <form>

    <div class="form-group row">
        <label for="qtlseqr_min_depth" class="col-xs-5" style="font-weight:normal;">Minimum Read Depth
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="To filter out low coverage/depth variations. <b>Default 5</b>. Any variation coverage less than the stetted value will be ignored. "
            >?</span
          >                
        </label>
        <div class="col-xs-7">
          <input id="qtlseqr_min_depth" type="input" value="" placeholder="5"></input></td>
        </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Population Structure
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="<b>F2</b>: Dominant character bulk data from F2 population is heterozygosity.
        </br>
        <b>RIL/DH</b>: Population genotypes of candidate gene are expected homozygote in both tow bulks.
            "
            >?</span
          >               
        </label>
        <div class="col-xs-7">
            <label class="radio-inline">
                <input type="radio" name="pop_stru" id="popStruc1" value="RIL" checked> RIL
              </label>
              <label class="radio-inline">
                <input type="radio" name="pop_stru" id="popStruc2" value="F2"> F2
              </label>               
        </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Ref Allele Frequency
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="Total ref allele coverage of two bulk / total coverage (both refer and allele) of two bulk. <b>Default 0.3</b>. Suggest 0.1~0.5. "
            >?</span
          >             
        </label>
        <div class="col-xs-7">
            <input id="qtlseqr_ref_freq" type="input" value="" placeholder="0.3" >          
        </div>
    </div>
    

  </form>
  <div class="panel-body">
    <input class="btn btn-primary btn-sm" id="submit" type="submit" value="Submit">
  </div>

</div>
{% endblock %}

{% block js %}
<script src="{{ url_for('static', filename='js/utils/bsa.js') }}"></script>
{% endblock %}
