{% extends 'base.html' %}
{% block css %}
<link
  href="{{ url_for('static', filename='css/tooltip.css') }}"
  rel="stylesheet"
/>
{% endblock %}
{% block content %}
<div class="row">
  <h2>Gene Mapping Via BSA(Bulked Segregant Analysis)</h2>
  <div class="panel-heading">
    <p>1. Wild Bulk and Mutant Bulk are <b>required</b>.</p>
    <p>2. Wild Parent and Mutant Parent are suggested if available.</p>
    <p>3. Background could help you filter out some variants, but you should be much careful the genetic
      background and phenotype of sample you selected.</p>
    <h3 class="panel-title">Please login and upload your vcf file if you not find the sample you want in the boxes
      below.</h3>
  </div></br>

  <div class="row">

    {% if pri_samples %}
    <div class="col-xs-3">
      <h3 class="panel-title">My samples</h3>
      <select name="from[]" id="multi_d" class="form-control multi_d" size="67" multiple="multiple">
        {% for sample in pri_samples %}
        <option value="private">{{ sample }}</option>
        {% endfor %}
      </select>
    </div>
    {% endif %}

    {% if pub_samples %}
    <div class="col-xs-3">
      <h3 class="panel-title">Public samples</h3>
      <select name="from[]" id="multi_d2" class="form-control multi_d" size="67" multiple="multiple">
        {% for sample in pub_samples %}
        <option value="public">{{ sample }}</option>
        {% endfor %}
      </select>
    </div>
    {% endif %}


    <div class="col-xs-2">
      <button type="button" id="multi_d_rightAll" class="btn btn-default btn-block" style="margin-top: 20px;"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>

      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_2" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>

      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_3" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>

      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_4" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>
      <hr style="margin: 40px 0 60px;" />

      <button type="button" id="multi_d_rightAll_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-forward"></i>&nbsp;ADD ALL</button>
      <button type="button" id="multi_d_rightSelected_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-right"></i>&nbsp;ADD</button>
      <button type="button" id="multi_d_leftSelected_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-chevron-left"></i>&nbsp;REMOVE</button>
      <button type="button" id="multi_d_leftAll_5" class="btn btn-default btn-block"><i
          class="glyphicon glyphicon-backward"></i>&nbsp;REMOVE ALL</button>
    </div>

    <div class="col-xs-4">
      <b>Wild Bulk (Required)</b>
      <select name="to[]" id="multi_d_to" class="form-control" size="8" multiple="multiple"></select>

      <br />
      <hr /><br />

      <b>Mutant Bulk (Required)</b>
      <select name="to_2[]" id="multi_d_to_2" class="form-control" size="8" multiple="multiple"></select>
      <br />
      <hr /><br />

      <b>Wild Parent (Optional)</b>
      <select name="to_3[]" id="multi_d_to_3" class="form-control" size="8" multiple="multiple"></select>
      <br />
      <hr /><br />

      <b>Mutant Parent (Optional)</b>
      <select name="to_4[]" id="multi_d_to_4" class="form-control" size="8" multiple="multiple"></select>
      <br />
      <hr /><br />

      <b>Background (Optional)</b>
      <select name="to_5[]" id="multi_d_to_5" class="form-control" size="8" multiple="multiple"></select>
    </div>
  </div>
  <br /><br />
  <h3 class="panel-title"><b>Optional parameters [SNP Density/varBScore]:</b></h3>
  <br>

  <form>

    <div class="form-group row">
        <label for="min_depth" class="col-xs-5" style="font-weight:normal;">Minimum Read Depth
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="To filter out low coverage/depth variations. <b>Default 5</b>. Any variation coverage less than 5 will be ignored."
            >?</span
          ></label>
        <div class="col-xs-7">
          <input id="min_depth" type="input" value="" placeholder="5" >
        </div>
    </div>

    <div class="form-group row">
      <label for="w-snp-num" class="col-xs-5" style="font-weight:normal;">SNP Window Size
        <span
        class="badge"
        data-toggle="tooltip"
        data-placement="top"
        data-html="true"
        title="This refers to the number of SNPs in a sliding window. <b>Default 20</b>. A size of 5-30 is suggested. The algorithm will calculate the SNP linkage-coefficient within a window. A smaller value will give a better resolution, and more SNPs in a given window will decrease more noise. Different parameters should be tested to obtain the best results."
        >?</span
      >
    </label>
      <div class="col-xs-7">
        <input id="w-snp-num" type="input" value="" placeholder="20" >
      </div>
    </div>

    <div class="form-group row">
      <label for="s-snp-num" class="col-xs-5" style="font-weight:normal;">SNP Step Size
        <span
        class="badge"
        data-toggle="tooltip"
        data-placement="top"
        data-html="true"
        title="This refers to the number of SNPs in a sliding step. <b>Default 5</b>. A size of 5-20 is suggested. A smaller value will give better resolution, but will take more time to calculate. When the data originate from an advanced generation population, it is possible to increase the number of SNPs step size appropriately. The default of 5 works well for most datasets."
        >?</span
      >    
    </label>
      <div class="col-xs-7">
        <input id="s-snp-num" type="input" value="" placeholder="5" >
      </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Allele Frequency Filter [<b>Wild/Mutant</b>]
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="This parameter will filter out SNPs with a high heterozygosity rate. We suggest 0.39 for F2 populations and 0.3 for F3/DH/RIL populations.
</br>
<b>F2 population</b>: The target genotype of the dominant-character bulk data from the F2 population is heterozygous, and the candidate gene allele frequency in the dominant-character bulk is expected to be 1/3 (AA:Aa=1:2). Thus, the parameter should be set to 0.36/0.39/0.43 or any value higher than 1/3 and lower than 0.5 for F2 populations. If the phenotype of all F2 plants has been validated in F3 and shown not to segregate, then the two bulk genotypes are expected to be homozygous, and the parameter should follow the rule as below.
<b>F3/RIL/DH populations</b>: The genotype at the target locus is expected to be homozygous in both bulks. The allele frequency should be equal or very close to 0 or 1. This parameter should be set to 0.1/0.2/0.3 or any value lower than 0.3.
            "
            >?</span
          >               
        </label>
        <div class="col-xs-7">
            <input id="ref_freq" type="input" value="" placeholder="0.3" >
        </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Allele Frequency Filter [<b>Wild/Mutant Parent</b>]
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="If the parental data are provided, this parameter will filter out SNPs with high heterozygosity. Usually, most alleles in the parental genotype are homozygous. The allele frequency is equal or very close to 0 or 1. We suggest setting this value to 0.05 or 0.1 to avoid the effect of sequencing errors. <b>Default 0.3</b>"
            >?</span
          >                
        </label>
        <div class="col-xs-7">
            <input id="p_ref_freq" type="input" value="" placeholder="0.3" >    
        </div>
    </div>    
    
    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Allele Frequency Filter [<b>Background</b>]
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="If a dataset is added to the background to help improve gene mapping, much care should be taken regarding the genetic backgrounds and phenotypes of the samples. The expected allele frequency of samples in this bulk will depend on which samples are selected."
            >?</span
          >          
        </label>
        <div class="col-xs-7">
            <input id="background_ref_freq" type="input" value="" placeholder="0.3" >       
        </div>
    </div>  
  </form>
  <br />
  <h3 class="panel-title"><b>Optional parameters [QTLseqr/ED/G']:</b></h3>
  <br>
  <form>

    <div class="form-group row">
        <label for="qtlseqr_min_depth" class="col-xs-5" style="font-weight:normal;">Minimum Read Depth
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="To filter out low coverage/depth variations. <b>Default 5</b>. Any variation coverage less than 5 will be ignored. "
            >?</span
          >                
        </label>
        <div class="col-xs-7">
          <input id="qtlseqr_min_depth" type="input" value="" placeholder="5"></input></td>
        </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Population Structure
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="<b>F2</b>: Dominant character bulk data from F2 population is heterozygosity.
        </br>
        <b>RIL/DH</b>: Population genotypes of candidate gene are expected homozygote in both tow bulks.
            "
            >?</span
          >               
        </label>
        <div class="col-xs-7">
            <label class="radio-inline">
                <input type="radio" name="pop_stru" id="popStruc1" value="RIL" checked> RIL/DH/F3/Homozygous Plants Bulk
              </label>
              <label class="radio-inline">
                <input type="radio" name="pop_stru" id="popStruc2" value="F2"> F2/Heterozygous Plants Bulk
              </label>               
        </div>
    </div>

    <div class="form-group row">
        <label class="col-xs-5" style="font-weight:normal;">Ref Allele Frequency
            <span
            class="badge"
            data-toggle="tooltip"
            data-placement="top"
            data-html="true"
            title="This refers to the total reference allele coverage of the two bulks/total coverage (both reference and allele) of the two bulks. The default is 0.3, and a value of 0.1~0.5 is suggested."
            >?</span
          >             
        </label>
        <div class="col-xs-7">
            <input id="qtlseqr_ref_freq" type="input" value="" placeholder="0.3" >          
        </div>
    </div>
    

  </form>
  <div class="panel-body">
    <input class="btn btn-primary btn-sm" id="submit" type="submit" value="Submit">
  </div>

</div>
{% endblock %}

{% block js %}
<script src="{{ url_for('static', filename='js/utils/bsa.js') }}"></script>
{% endblock %}
