{% extends 'base.html' %} {% block content %}
<div class="title" style="margin-bottom: 30px;">
  <h3>Citation</h3>
  </br>
<h4>WheatGmap</h4>
<p>Zhang, L., Dong, C., Chen, Z., Gui, L., Chen, C., Li, D., Xie, Z., Zhang, Q., Zhang, X., Xia, C., Liu, X., Kong, X., & Wang, J. (2020). WheatGmap: A Comprehensive Platform for Wheat Gene Mapping and Genomic Studies. Molecular Plant. https://doi.org/https://doi.org/10.1016/j.molp.2020.11.018</p> 
  </br>
  <h4>Genome and Gene annotation</h4>
  <p>Appels, R., Eversole, K., Feuillet, C., Keller, B., Rogers, J., Stein, N., Pozniak, C. J., Choulet, F., Distelfeld, A., Poland, J., Ronen, G., Barad, O., Baruch, K., Keeble-Gagnère, G., Mascher, M., Ben-Zvi, G., Josselin, A. A., Himmelbach, A., Balfourier, F., … Wang, L. (2018). Shifting the limits in wheat research and breeding using a fully annotated reference genome. Science, 361(6403). https://doi.org/10.1126/science.aar7191</p>
  <p>Ramírez-González, R. H., Borrill, P., Lang, D., Harrington, S. A., Brinton, J., Venturini, L., Davey, M., Jacobs, J., Van Ex, F., Pasha, A., Khedikar, Y., Robinson, S. J., Cory, A. T., Florio, T., Concia, L., Juery, C., Schoonbeek, H., Steuernagel, B., Xiang, D., … Uauy, C. (2018). The transcriptional landscape of polyploid wheat. Science, 361(6403). https://doi.org/10.1126/science.aar6089</p>
  </br>
  <h4>BSA mapping</h4>
<p><b>SNP-index:</b></p>
<p>Abe, A., Kosugi, S., Yoshida, K., Natsume, S., Takagi, H., Kanzaki, H., Matsumura, H., Yoshida, K., Mitsuoka, C., Tamiru, M., Innan, H., Cano, L., Kamoun, S., & Terauchi, R. (2012). Genome sequencing reveals agronomically important loci in rice using MutMap. Nature Biotechnology, 30(2). https://doi.org/10.1038/nbt.2095</p>

<p>Mansfeld, B. N., & Grumet, R. (2018). QTLseqr: An R Package for Bulk Segregant Analysis with Next-Generation Sequencing. The Plant Genome, 11(2), 0. https://doi.org/10.3835/plantgenome2018.01.0006</p>

<p><b>varBScore:</b></p>
<p>Dong, C., Zhang, L., Chen, Z., Xia, C., & Gu, Y. (2020). Combining a New Exome Capture Panel With an Effective varBScore Algorithm Accelerates BSA-Based Gene Cloning in Wheat. 11(August), 1–12. https://doi.org/10.3389/fpls.2020.01249</p>

<p><b>ED:</b></p>
<p>Gorsi, B., Yost, H. J., Demarest, B. L., Bisgrove, B. W., Hill, J. T., & Su, Y.-C. (2013). MMAPPR: Mutation Mapping Analysis Pipeline for Pooled RNA-seq. Genome Research, 23(4), 687–697. https://doi.org/10.1101/gr.146936.112</p>

<p><b>G prime:</b></p>
<p>Gorsi, B., Yost, H. J., Demarest, B. L., Bisgrove, B. W., Hill, J. T., & Su, Y.-C. (2013). MMAPPR: Mutation Mapping Analysis Pipeline for Pooled RNA-seq. Genome Research, 23(4), 687–697. https://doi.org/10.1101/gr.146936.112</p>

<p><b>QTLseqr:</b></p>
<p>Mansfeld, B. N., & Grumet, R. (2018). QTLseqr: An R Package for Bulk Segregant Analysis with Next-Generation Sequencing. The Plant Genome, 11(2), 0. https://doi.org/10.3835/plantgenome2018.01.0006</p>
  
</br>
  <h4>BSA Demo</h4>
  <p>Harrington, S. A., Cobo, N., Karafiátová, M., Doležel, J., Borrill, P., & Uauy, C. (2019). Identification of a Dominant Chlorosis Phenotype Through a Forward Screen of the Triticum turgidum cv. Kronos TILLING Population. Frontiers in Plant Science, 10. https://doi.org/10.3389/fpls.2019.00963</p>
  <p>Gardiner, L.-J., Bansept-Basler, P., El-Soda, M., Hall, A., & O’Sullivan, D. M. (2019). A framework for gene mapping in wheat demonstrated using the Yr7 yellow rust resistance gene. BioRxiv, 793604. https://doi.org/10.1101/793604</p>
<p>Dong, C., Zhang, L., Chen, Z., Xia, C., & Gu, Y. (2020). Combining a New Exome Capture Panel With an Effective varBScore Algorithm Accelerates BSA-Based Gene Cloning in Wheat. 11(August), 1–12. https://doi.org/10.3389/fpls.2020.01249</p>

  <h4></h4>

</div>
{% endblock %}
