{% extends 'base.html' %} 
{% block css %}
<link
  href="{{ url_for('static', filename='css/atom-one-dark.css') }}"
  rel="stylesheet"
/>
{% endblock %}
{% block content %}
<div class="title" style="margin-bottom: 30px;">
  <h3>VCF Prepare</h3>
</br>
  <h4>Required</h4>
  <p>> The <b>reference genome</b> for variant-calling must be <a href="https://www.wheatgenome.org/" target="_blank"><b>IWGSC RefSeq V1.0/V1.1</b></a> at this time.</p>
  <p>> VCF can work out flow by <a href="https://gatk.broadinstitute.org/" target="_blank">GATK Best Practice</a> or
    <a href="https://samtools.github.io/bcftools/howtos/variant-calling.html" target="_blank">Samtools</a>
    variant-calling pipeline with both DNA and RNA data.</p>
  <p>> All VCF need be compressed as <b>VCF.gz</b> by bcftools or GATK.</p>
  <p>> VCF should contain “GT,AD” in FORMAT tags. </p>
  <p>> VCF from GATK pipeline with default perameters already has "GT,AD" in INFO column as the first and second tags.
  </p>
  <p>> If you call variants with samtools/bcftools, you need specify the output with “FORMAT/AD”</p>
  <p><B>Joint-Call Cohort is suggested for callling variants VCF.</B></p>
  <p>Here are two demo for cohort variants-calling below:</p>
  </br>
  <p><B>GATK</B></p>
  <pre style="padding: 0;"><code>
    gatk HaplotypeCaller \
        --input mutant.bam \
        --output mutant.g.vcf.gz \
        --reference refer \
        --emit-ref-confidence GVCF
    
    # wild gvcf
    gatk HaplotypeCaller \
        --input wild.bam \
        --output wild.g.vcf.gz \
        --reference refer \
        --emit-ref-confidence GVCF
    
    # merge gvcf
    gatk CombineGVCFs \
        --output bsa.g.vcf.gz \
        --reference $refer \
        --variant mutant.g.vcf.gz \
        --variant wild.g.vcf.gz
    
    # call vcf
    gatk GenotypeGVCFs \
        --reference $refer \
        --variant bsa.g.vcf.gz \
        --output bsa.vcf.gz      
  </code></pre>
  </div>
  </br>
  </br>
  <p><B>Samtools/Bcftools</B></p>
  <pre style="padding: 0;"><code>
    bcftools mpileup -Ou --annotate "FORMAT/AD" \
        -f ref.fa aln.mutant_bulk.bam aln.wild_bulk.bam | \
            bcftools call -Ou -mv | \
                bcftools filter -Oz -s LowQual -e '%QUAL<20 || DP>100' > bsa.vcf.gz      
  </code></pre>
  <p><a href="http://www.htslib.org/workflow/" target="_blank">read more</a> </p>

</div>
{% endblock %}
{% block js %}
<script src="https://cdn.bootcdn.net/ajax/libs/highlight.js/10.0.0/highlight.min.js"></script>
<script src="{{ url_for('static', filename='js/utils/initHightlight.js') }}"></script>
{% endblock %}
