{% extends 'base.html' %} {% block content %}
<div class="title" style="margin-bottom: 30px;">
  <h3>VCF Prepare</h3>
</br>
  <h4>Required</h4>
  <p>> The <b>reference genome</b> for variant-calling must be <a href="www.wheatgenome.org
    " target="_blank"><b>IWGSC RefSeq V1.0</b></a> at this time.</p>
  <p>> VCF can work out flow by <a href="https://gatk.broadinstitute.org/" target="_blank">GATK Best Practice</a> or
    <a href="https://samtools.github.io/bcftools/howtos/variant-calling.html" target="_blank">Samtools</a>
    variant-calling pipeline with both DNA and RNA data.</p>
  <p>> All VCF need be compressed as <b>VCF.gz</b> by bcftools or GATK.</p>
  <p>> VCF should contain “GT,AD” in FORMAT tags. </p>
  <p>>> VCF from GATK pipeline with default paramets already has "GT,AD" in INFO colum as the first and sencond tags.
  </p>
  <p>>> If you call variants with samtools/bcftools, you need specify the output with “FORMAT/AD”</p>
  </br>
  <h4>Optional</h4>
  <p><B>Joint-Call Cohort is suggested for callling variants VCF.</B></p>
  <p>Here are some demo for cohort variants-calling below:</p>
  </br>
  <p><B>GATK</B></p>
  <div type="margin-left:20%">
    <p># mutant gvcf</br>
      gatk HaplotypeCaller \</br>
      --input mutant.bam \</br>
      --output mutant.g.vcf.gz \</br>
      --reference refer \</br>
      --emit-ref-confidence GVCF</br>
      </br>
      # wild gvcf</br>
      gatk HaplotypeCaller \</br>
      --input wild.bam \</br>
      --output wild.g.vcf.gz \</br>
      --reference refer \</br>
      --emit-ref-confidence GVCF</br>
      </br>
      # merge gvcf
      gatk CombineGVCFs \</br>
      --output bsa.g.vcf.gz \</br>
      --reference $refer \</br>
      --variant mutant.g.vcf.gz \</br>
      --variant wild.g.vcf.gz</br>
      </br>
      # call vcf
      gatk GenotypeGVCFs \</br>
      --reference $refer \\/br>
      --variant bsa.g.vcf.gz \</br>
      --output bsa.vcf.gz</br>
    </p>
  </div>
  </br>
  </br>
  <p><B>Samtools/Bcftools</B></p>
  <p>
    bcftools mpileup -Ou --annotate "FORMAT/AD" \</br>
    -f ref.fa aln.mutant_bulk.bam aln.wild_bulk.bam | \</br>
    bcftools call -Ou -mv | \</br>
    bcftools filter -Oz -s LowQual -e '%QUAL<20 || DP>100' > bsa.vcf.gz
  </p>

</div>
{% endblock %}